integrations/object_store/src/store.rs

```diff
@@ -412,6 +417,74 @@ impl ObjectStore for OpendalStore {
     }
 }
 
+/// `MultipartUpload`'s impl based on `Writer` in opendal
+struct OpendalUpload {
+    // Refer to the fs impl in `object_store`:
+    // https://github.com/apache/arrow-rs/blob/a35214f92ad7c3bce19875bb091cb776447aa49e/object_store/src/local.rs#L715
+    // IO exists during locking, so use tokio::sync::Mutex here
+    writer: Arc<Mutex<Writer>>,
+
+    location: Path,
+}
+
+impl OpendalUpload {
+    fn new(writer: Writer, location: Path) -> Self {
+        Self {
+            writer: Arc::new(Mutex::new(writer)),
+            location,
+        }
+    }
+}
+
+#[async_trait]
+impl MultipartUpload for OpendalUpload {
+    fn put_part(&mut self, data: PutPayload) -> UploadPart {
+        let writer = self.writer.clone();
+        let location = self.location.clone();
+
+        async move {
+            let mut writer = writer.lock().await;
+            writer
+                .write(Buffer::from_iter(data.into_iter()))
+                .await
+                .map_err(|err| format_object_store_error(err, location.as_ref()))
+        }
+        .into_send()
+        .boxed()
+    }
+
+    async fn complete(&mut self) -> object_store::Result<PutResult> {
+        let mut writer = self.writer.lock().await;
+        writer
+            .close()
+            .into_send()
+            .await
+            .map_err(|err| format_object_store_error(err, self.location.as_ref()))?;
+
+        Ok(PutResult {
+            e_tag: None,
+            version: None,
+        })
+    }
+
+    async fn abort(&mut self) -> object_store::Result<()> {
+        let mut writer = self.writer.lock().await;
+        writer
+            .abort()
+            .into_send()
+            .await
+            .map_err(|err| format_object_store_error(err, self.location.as_ref()))
+    }
+}
+
+impl fmt::Debug for OpendalUpload {
```

Comment:
`Debug` has been included, so we can remove the usage of `fmt`. Or we can remove all `Debug` / `Formatter` instead.

Reply:
Done